src/themis/secure_cell_seal_passphrase.c

```diff
@@ -0,0 +1,341 @@
+                                                                 size_t* encrypted_message_length)
+{
+    themis_status_t res = THEMIS_FAIL;
+    uint8_t iv[THEMIS_AUTH_SYM_IV_LENGTH] = {0};
+    uint8_t salt[THEMIS_AUTH_SYM_PBKDF2_SALT_LENGTH] = {0};
+    uint8_t auth_tag[THEMIS_AUTH_SYM_AUTH_TAG_LENGTH] = {0};
+    uint8_t derived_key[THEMIS_AUTH_SYM_KEY_LENGTH / 8] = {0};
+    struct themis_scell_auth_token_passphrase hdr = {0};
+    struct themis_scell_pbkdf2_context kdf = {0};
+    size_t auth_tag_length = 0;
+
+    /* Message length is currently stored as 32-bit integer, sorry */
+    if (message_length > UINT32_MAX) {
+        return THEMIS_INVALID_PARAMETER;
+    }
+
+    kdf.iteration_count = THEMIS_AUTH_SYM_PBKDF2_ITERATIONS;
+    kdf.salt = salt;
+    kdf.salt_length = sizeof(salt);
+
+    hdr.alg = THEMIS_AUTH_SYM_PASSPHRASE_ALG;
+    hdr.iv = iv;
+    hdr.iv_length = sizeof(iv);
+    hdr.auth_tag = auth_tag;
+    hdr.auth_tag_length = sizeof(auth_tag);
+    hdr.message_length = (uint32_t)message_length;
+    /* KDF pointer is ignored but size is important */
+    hdr.kdf_context = NULL;
+    hdr.kdf_context_length = (uint32_t)themis_scell_pbkdf2_context_size(&kdf);
+
+    res = soter_rand(salt, sizeof(salt));
+    if (res != THEMIS_SUCCESS) {
+        goto error;
+    }
+
+    res = soter_pbkdf2_sha256(passphrase,
+                              passphrase_length,
+                              kdf.salt,
+                              kdf.salt_length,
+                              kdf.iteration_count,
+                              derived_key,
+                              sizeof(derived_key));
+    if (res != THEMIS_SUCCESS) {
+        goto error;
+    }
+
+    res = soter_rand(iv, sizeof(iv));
+    if (res != THEMIS_SUCCESS) {
+        goto error;
+    }
+
+    /* We are doing KDF ourselves, ask Soter to not interfere */
+    auth_tag_length = hdr.auth_tag_length;
+    res = themis_auth_sym_plain_encrypt(soter_alg_without_kdf(hdr.alg),
+                                        derived_key,
+                                        sizeof(derived_key),
+                                        hdr.iv,
+                                        hdr.iv_length,
+                                        user_context,
+                                        user_context_length,
+                                        message,
+                                        message_length,
+                                        encrypted_message,
+                                        encrypted_message_length,
+                                        auth_tag,
+                                        &auth_tag_length);
+    if (res != THEMIS_SUCCESS) {
+        goto error;
+    }
+    /*
+     * We should have allocated just the right amount here, but if our guess
+     * was off then the header is not usable due to bothched data offsets.
+     */
+    if (auth_tag_length != hdr.auth_tag_length) {
```

Comment:
> can you explain in which cases [auth token length] can be not equal?

Normally it shouldn't. The only case which I envision is _our_ programming error when we introduce a new algorithm with a different auth tag length, or accidentally change the tag length constant. So this condition _should_ never fail. But I figured that we'd better leave a failsafe here, just in case, since we pass this function an out-parameter, it might change it, and some future code changed might affect that.

Reply:
1. if someone add one more algrorithm, it will not be used anywhere without changing constants/defines or rewriting code.
2. all static errors better to check in tests or on compile time to reduce runtime overhead. it's static error which can be tested. I though that maybe I missed some case with runtime error with user input, but auth tag and length we generate and return, and it's not depends on any user's input and configured at compile time. I personally prefer to keep functions simple and fast as it possible, and let such checks for compilator and tests )

plus I don't understand why we need to store auth_tag_length in two places:
1. `hdr.auth_tag_length = sizeof(auth_tag);`
2. auth_tag_length = hdr.auth_tag_length;
3. use `auth_tag_length` in `themis_auth_sym_plain_encrypt` call
4. compare them `auth_tag_length != hdr.auth_tag_length` first time
5. check them second time `*auth_token_length < themis_scell_auth_token_passphrase_size(&hdr)`. just to be sure that in `themis_auth_sym_plain_encrypt` call it wasn't changed?

Reply:
> I don't understand why we need to store auth_tag_length in two places

We have to use a temporary variable because of some interface impedance:

```c
struct themis_scell_auth_token_passphrase {
    // ...
    const uint8_t* auth_tag;
    uint32_t auth_tag_length;
    // ...
};
```

```c
themis_status_t themis_auth_sym_plain_encrypt(uint32_t alg,
                                              // ...
                                              uint8_t* auth_tag,
                                              size_t* auth_tag_length)
```

Since themis_auth_sym_plain_encrypt() accepts a pointer to size_t, we can't pass uint32_t* there. And it accepts size_t* because soter_sym_aead_encrypt_final() accepts size_t*.

Secure Cell stores auth tag length as a 32-bit value but that's an arbitrary decision of Secure Cell, we should not change Soter interface because of that. However, just now it had occured to me that we probably _can_ change themis_auth_sym_plain_encrypt() to work with uint32_t since that's our internal helper, and it we would be able to avoid having this temporary here. It will still be there inside themis_auth_sym_plain_encrypt(), along with a precautionary range check I guess.

Themis currently hardcodes the length of authentication tag because we ‘just know’ what size it will have for the algorithms we use, but Soter interface assumes that the length of the tag is output parameter which should be measured by the caller, just like with any other of our buffer API. Maybe we should do that instead...

Anyway, I'll see what can be done here to make the code cleaner.

Reply:
> Soter interface assumes that the length of the tag is output parameter which should be measured by the caller, just like with any other of our buffer API. Maybe we should do that instead...

hm, I thought that value for tag_length is constant in themis module which depends on soter's constant
but I see that it is not - https://github.com/cossacklabs/themis/blob/3c8bf73778972c6d31d926313b2095722e4e0764/src/themis/secure_cell_alg.h#L24

yes, it's will be the most correct way to deal with lengths. it will allow to avoid checks of relations between themis constants and soter internal implementations

Reply:
I've looked at the “measuring” angle but it turned out to be really inconvenient. So I've tweaked the type of internal helper function and this allowed to get rid of this strange error check here for passphrase API and a similar one in master key API.

We still have a check in a different place, but that one should always succeed. And hey, it's still –8 lines!